### DISCUSSION TITLE
Seeking for advice on ImageMagick command line

### DISCUSSION THREAD

**Discussion by satimis** [ImageMagick/ImageMagick#5197] [Digital Image Processing]
Hi all
Converting negative images scanned on film negatives with mobile phone

On GIMP I ran following steps
Colors -> Auto -> White Balance
then;
Colors -> Invert or Linear Invert
Done

Please advice what will be the command line on ImageMagick to get the job done?  Thanks
Regards

**Comment by fmw42** [comment]
On Imagemagick 7,

`magick image.suffix -auto-level -white-balance -negate result.suffix`

where image.suffix is your input image and result.suffix is your output (suffix is your image type, e.g. jpg, png, etc)

Please always provide your Imagemagick version and OS when asking questions.

**Comment by satimis** [comment]
Thanks for your advice.

OS - Ubuntu 20.04 desktop

$ convert --version
```
Version: ImageMagick 6.9.10-23 Q16 x86_64 20190101 https://imagemagick.org
Copyright: © 1999-2019 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC Modules OpenMP 
Delegates (built-in): bzlib djvu fftw fontconfig freetype jbig jng jpeg lcms lqr ltdl lzma openexr pangocairo png tiff webp wmf x xml zlib
```
'magick" is replaced with "convert" here

$ convert 20220606_a1.jpg -auto-level -white-balance -negate 20220606_a1_output.jpg
`convert-im6.q16: unrecognized option `-white-balance' @ error/convert.c/ConvertImageCommand/3240.`

Regards

**Reply by fmw42** [reply]
IM 6 does not have -white-balance.  So you cannot use convert. You would have to upgrade to IM 7 to have -white-balance. But you can do the rest with IM 6 convert

`convert image.suffix -auto-level -negate result.suffix`

See my scripts, whitebalancing, neg2pos and negative2positive at http://www.fmwconcepts.com/imagemagick for doing that in IM 6.

**Comment by satimis** [comment]
Hi,
Hi fmw42

Thanks for your advice.
Reinstall ImageMagick

$ identify -version
`Version: ImageMagick 7.1.0-37 Q16-HDRI x86_64 20104 https://imagemagick.org`

$ magick p_01_20220523.jpeg -auto-level -white-balance -negate p_01_20220523_positive.jpeg
```
magick: unable to open image 'p_01_20220523.jpeg': No such file or directory @ error/blob.c/OpenBlob/3533.
magick: no decode delegate for this image format `JPEG' @ error/constitute.c/ReadImage/738.
```
Still fail.  Please advise how to fix the problem. Thanks

Regards

**Reply by fmw42** [reply]
How did you install Imagemagick? Did you do it from a binary or from source. If from source, then you have to pre-install all the desired delegate libraries before compiling Imagemagick.

What do you get from

`magick -version`

Is your p_01_20220523.jpeg image in your working directory. If not, then either cd to that directory or put the full or relative path to the image in the file name.

**Reply by satimis** [reply]
$ magick -version
```
Version: ImageMagick 7.1.0-37 Q16-HDRI x86_64 20104 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(4.5) 
Delegates (built-in): fontconfig freetype jp2 ltdl png webp x zlib
Compiler: gcc (9.4)
```
I followed below document to install ImageMagick
How to install ImageMagick 7 on Ubuntu
https://scottiestech.info/2020/05/26/how-to-install-imagemagick-7-on-ubuntu/comment-page-1/

p_01_20220523.jpeg is on Desktop.  I copied it there for testing.

~/Desktop$ ls`
`20220606_a1.jpg  p_01_20220523.jpeg  PC1_20220604  PC1a_20220604``

I also tested 20220606_a1.jpg. and tried "convert"  The result is the same

Regards

**Reply by fmw42** [reply]
You show:

`Delegates (built-in): fontconfig freetype jp2 ltdl png webp x zlib`

So you do not have any delegate for JPEG. You would need to include that delegate when compiling.

**Reply by satimis** [reply]
Thanks for your advice.

I followed below document to install IM7 from source.  Please advise how to do that?
How to install ImageMagick 7 on Ubuntu
https://scottiestech.info/2020/05/26/how-to-install-imagemagick-7-on-ubuntu/comment-page-1/

Besides how about reading .pdf files?

Thanks

Regards

**Reply by fmw42** [reply]
Sorry, I am not an expert on installation especially not for Linux.  You can read PDF if you install Ghostscript.

**Reply by satimis** [reply]
Now I can run IM7 without problem after installing jpegsrc.v9e.tar.gz.

I expect learning something new to me and am prepared to go another round installing IM7 on Ubuntu 20.04

How To Install ghostscript on Ubuntu 20.04
https://installati.one/ubuntu/20.04/ghostscript/

Whether I need to install Ghostscript to running IM7 ?

Regards

**Reply by fmw42** [reply]
You only need Ghostscript to permit Imagemagick to read PDF files. It is not needed to write PDF files. But writing PDF simply inserts the raster image into a vector shell. It does not convert the raster image to vectors data.

**Comment by snibgo** [comment]
The list of built-in delegates doesn't include jpeg. I expect that explains why IM can't read jpeg files.

**Reply by satimis** [reply]
I have 2 image files on Desktop;

20220606_a1.jpg
p_01_20220523.jpeg

I copied them on Desktop for testing.
:~/Desktop$ convert -rotate 90 20220606_a1.jpg 20220606_a1_90.jpg 
```
convert: no decode delegate for this image format `JPG' @ error/constitute.c/ReadImage/738.
convert: no images defined `20220606_a1_90.jpg' @ error/convert.c/ConvertImageCommand/3325.

```
The same error

Again
~/Desktop$ mogrify -rotate 90 20220606_a1.jpg 
`mogrify: no decode delegate for this image format `JPG' @ error/constitute.c/ReadImage/738.`

I'm very surprising.  Both convert and mogrify work seamlessly on ImageMagick 6 which was installed on Ubuntu 20.04 repo.  ImageMagick 7 was built on source.  Have I missed something?

Regards

**Comment by snibgo** [comment]
>  Have I missed something?

> no decode delegate for this image format `JPG'

You have built IM without the built-in delegate to decode JPG files.

**Reply by fmw42** [reply]
Perhaps this will help.  https://askubuntu.com/questions/1042436/how-to-install-delegate-libraries-for-image-magick-7-0-7

**Reply by satimis** [reply]
Hi all,

Thanks for your reply.

I have tried following Step-2 of my linked article.

Step 2: Just in case, install all the build dependencies for ImageMagick:

$ sudo apt build-dep imagemagick
```
Reading package lists... Done
E: You must put some 'deb-src' URIs in your sources.list
```
I couldn't figure it out how to solve it and skipped it.

Regards

**Comment by satimis** [comment]
Hi all,

Problem solved after download **jpegsrc.v9e.tar.gz** and install it on the same folder of **ImageMagick 7**

In my case
**/usr/local/src/**

Now I can run;
$ convert input.jpg -auto-level -white-balance -negate output.jpg

without complaint.  **magick** is the same.

**Download jpegsrc.v9e.tar.gz**
Independent JPEG Group
http://www.ijg.org/

Furthermore, if I find the quality of the positive image not up to my satisfaction and I need;
Replace **"white balance"** with **"equalize"**
Replace **"invert"** with **"linear invert"**

What will be the command ?  Thanks

Regards

Edit
===
On executing fmw42's command;
`magick image.suffix -auto-level -white-balance -negate result.suffix`
I got the result as test_1.png (photo attached)

But I expect the result as test_2.png (photo attached).  What command shall I run?

On GIMP I only change "Invert" to "Linear Invert
![test_1_orig](https://user-images.githubusercontent.com/30752517/172837711-ef6e452b-b747-4283-83da-5eefeceb26cf.png)
![test_2_orig](https://user-images.githubusercontent.com/30752517/172837760-1a4a527c-bd5c-4d2b-8d99-01157a40de19.png)
"
